exp/bufferpool/pool.go

```diff
@@ -4,49 +4,161 @@ package bufferpool
 import "sync"
 
 const (
-	nBuckets = 20
+	defaultMinSize     = 1024
+	defaultBucketCount = 10
 )
 
-// A default global pool.
-var Default Pool
-
-// A pool of buffers, in variable sizes.
-type Pool struct {
-	buckets [nBuckets]sync.Pool
-}
-
-// Get a buffer of the given length. Its capacity may be larger than the
-// requested size.
-func (p *Pool) Get(size int) []byte {
-	for i := 0; i < nBuckets; i++ {
-		capacity := 1 << i
-		if capacity >= size {
-			var ret []byte
-			item := p.buckets[i].Get()
-			if item == nil {
-				ret = make([]byte, capacity)
-			} else {
-				ret = item.([]byte)
-			}
-			ret = ret[:size]
-			return ret
-		}
+var pool Pool = &BasicPool{}
+
+// Set the global pool.  This can be used to override the default implementation
+// to optimize for different allocation profiles. For example, applications that
+// target embedded platforms may wish to lower the minimum buffer size from 1KiB
+// to 512B, or less.
+//
+// The default implementation uses BasicPool, which is suitable for a wide range
+// of applications.  Most applications will not benefit from using a custom pool,
+// so be sure to profile.
+//
+// Set panics if p == nil.
+func Set(p Pool) {
+	if p == nil {
+		panic("Pool(nil)")
 	}
-	return make([]byte, size)
+
+	pool = p
+}
+
+// Get a buffer of with the specified size from the global pool.  The
+// resulting buffer's capacity MAY exceed size. The buffer SHOULD NOT
+// be resized beyond its underlying capacity, or it will leak to the
+// garbage collector.
+func Get(size int) []byte {
+	return pool.Get(size)[:size]
 }
 
-// Return a buffer to the pool. Zeros the slice (but not the full backing array)
-// before making it available for future use.
-func (p *Pool) Put(buf []byte) {
-	for i := 0; i < len(buf); i++ {
+// Put returns the buffer to the global pool.  The slice will be zeroed
+// before returning it to the pool.   Note that only the first len(buf)
+// bytes are zeroed, and not the full backing array.
+//
+// TODO:  should we zero out the entire array?  Users may have resized
+// the slice, so it may contain sensitive data.
+func Put(buf []byte) {
+	for i := range buf {
 		buf[i] = 0
 	}
 
-	capacity := cap(buf)
-	for i := 0; i < nBuckets; i++ {
-		if (1 << i) == capacity {
-			p.buckets[i].Put(buf[:capacity])
+	pool.Put(buf[:cap(buf)])
+}
+
+// Pool is a free list of []byte buffers of heterogeneous sizes.
+// Implementations MUST be thread safe and MUST support efficient
+// paging of buffers with arbitrary capacity.   Most applications
```

Comment:
I'm not exactly sure what paging means in this context.

Reply:
Fixed (dunno what I was thinking).